File: src/app/pages/materias/materia.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Materia {
  idMateria: number;
  nombre: string;
  costo: number;
}

@Injectable({ providedIn: 'root' })
export class MateriaService {
  private apiUrl = 'http://localhost:5242/api/materia';

  constructor(private http: HttpClient) {}

  insertarMateria(materia: Materia): Observable<any> {
    return this.http.post(`${this.apiUrl}/registrar`, materia);
  }

  obtenerMaterias(): Observable<Materia[]> {
    return this.http.get<Materia[]>(`${this.apiUrl}/obtener`);
  }

  buscarNombre(nombre: string): Observable<Materia> {
    return this.http.get<Materia>(`${this.apiUrl}/buscar/${nombre}`);
  }
  
  eliminarMateria(id: number): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`);
  }
  obtenerCostoTotal(idAlumno: number): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/costo-total/${idAlumno}`);
  }
  
  obtenerMateriasDeAlumno(idAlumno: number): Observable<Materia[]> {
    return this.http.get<Materia[]>(`${this.apiUrl}/materias-alumno/${idAlumno}`);
  }
  
  agregarMateriaAlumno(idAlumno: number, idMateria: number): Observable<any> {
    return this.http.post(`${this.apiUrl}/agregar-materia`, {
      idAlumno,
      idMateria
    });
  }
  

}

File: ControlEscolar.Services/MateriaService.cs
using ControlEscolar.Core.Models;
using ControlEscolar.Data;

namespace ControlEscolar.Services
{
    public class MateriaService
    {
        private readonly MateriaRepository _repository;

        public MateriaService(MateriaRepository repository)
        {
            _repository = repository;
        }

        public void RegistrarMateria(string nombre, decimal costo)
        {
            var materia = new Materia
            {
                Nombre = nombre,
                Costo = costo
            };
            _repository.InsertarMateria(materia);
        }

        public Materia? BuscarPorNombre(string nombre)
{
    return _repository.BuscarPorNombre(nombre);
}

public void EliminarMateria(int id)
{
    _repository.EliminarMateria(id);
}        public List<Materia> ObtenerMateriasDeAlumno(int idAlumno)
{
    return _repository.ObtenerMateriasDeAlumno(idAlumno).ToList();
}
public List<Materia> ObtenerMaterias()
{
    return _repository.ObtenerMaterias();
}

public decimal ObtenerCostoTotal(int idAlumno)
{
    return _repository.ObtenerCostoTotal(idAlumno);
}

public void AgregarMateriaAlumno(int idAlumno, int idMateria)
{
    _repository.AgregarMateriaAlumno(idAlumno, idMateria);
}
    }

}
